src/backend/distributed/utils/enable_ssl.c

```diff
@@ -0,0 +1,444 @@
+		configChanged = true;
+	}
+
+	if (configChanged)
+	{
+		GloballyReloadConfig();
+	}
+
+	PG_RETURN_NULL();
+}
+
+
+/*
+ * GloballyReloadConfig signals postmaster to reload the configuration as well as
+ * reloading the configuration in the current backend. By reloading the configuration in
+ * the current backend the changes will also be reflected in the current transaction.
+ */
+static void
+GloballyReloadConfig()
+{
+	if (kill(PostmasterPid, SIGHUP))
+	{
+		ereport(WARNING, (errmsg("failed to send signal to postmaster: %m")));
+	}
+	ProcessConfigFile(PGC_SIGHUP);
+}
+
+
+#ifdef USE_SSL
+
+/*
+ * EnsureReleaseOpenSSLResource registers the openssl allocated resource to be freed when the
+ * current memory context is reset.
+ */
+static void
+EnsureReleaseOpenSSLResource(MemoryContextCallbackFunction func, void *arg)
+{
+	MemoryContextCallback *cb = MemoryContextAllocZero(CurrentMemoryContext,
+													   sizeof(MemoryContextCallback));
+	cb->func = func;
+	cb->arg = arg;
+	MemoryContextRegisterResetCallback(CurrentMemoryContext, cb);
+}
+
+
+/*
+ * ShouldUseAutoSSL checks if citus should enable ssl based on the connection settings it
+ * uses for outward connections. When the outward connection is configured to require ssl
+ * it assumes the other nodes in the network have the same setting and therefor it will
+ * automatically enable ssl during installation.
+ */
+static bool
+ShouldUseAutoSSL(void)
+{
+	const char *sslmode = NULL;
+	sslmode = GetConnParam("sslmode");
+
+	if (strcmp(sslmode, "require") == 0)
+	{
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * CreateCertificatesWhenNeeded checks if the certificates exists. When they don't exist
+ * they will be created. The return value tells whether or not new certificates have been
+ * created. After this function it is guaranteed that certificates are in place. It is not
+ * guaranteed they have the right permissions as we will not touch the keys if they exist.
+ */
+static bool
+CreateCertificatesWhenNeeded()
+{
+	EVP_PKEY *privateKey = NULL;
+	X509 *certificate = NULL;
+	bool certificateWritten = false;
+	SSL_CTX *sslContext = NULL;
+
+	/*
+	 * Since postgres might not have initialized ssl at this point we need to initialize
+	 * it our self to be able to create a context. This code is less extensive then
+	 * postgres' initialization but that will happen when postgres reloads its
+	 * configuration with ssl enabled.
+	 */
+#ifdef HAVE_OPENSSL_INIT_SSL
+	OPENSSL_init_ssl(OPENSSL_INIT_LOAD_CONFIG, NULL);
+#else
+	SSL_library_init();
+#endif
+
+	sslContext = SSL_CTX_new(SSLv23_method());
+	if (!sslContext)
+	{
+		return false;
+	}
+	EnsureReleaseOpenSSLResource((MemoryContextCallbackFunction) & SSL_CTX_free,
+								 sslContext);
+
+	/*
+	 * check if we can load the certificate, when we can we assume the certificates are in
+	 * place. No need to create the certificates and we can exit the function
+	 */
+	if (SSL_CTX_use_certificate_chain_file(sslContext, ssl_cert_file) == 1)
```

Comment:
Worth noting in a comment (to anyone who dares change this) that this also makes the function idempotent, because if we fail without writing a valid certificate to disk (which is the last step) then we will retry the next time.